src/network/networkd-dhcp4.c

```diff
@@ -138,6 +139,151 @@ static bool link_prefixroute(Link *link) {
                 link->manager->dhcp4_prefix_root_cannot_set_table;
 }
 
+static int dhcp_set_routing_policy_rules(Link *link) {
+        const struct in_addr *router = NULL;
+        union in_addr_union address;
+        RoutingPolicyRule *rule;
+        struct in_addr netmask;
+        unsigned prefixlen;
+        int r;
+
+        assert(link);
+        assert(link->dhcp_lease);
+
+        r = sd_dhcp_lease_get_address(link->dhcp_lease, &address.in);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "DHCP error: no address: %m");
+
+        r = sd_dhcp_lease_get_netmask(link->dhcp_lease, &netmask);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "DHCP error: no netmask: %m");
+
+        prefixlen = in4_addr_netmask_to_prefixlen(&netmask);
+
+        (void) sd_dhcp_lease_get_router(link->dhcp_lease, &router);
+
+        if (DEBUG_LOGGING) {
+                _cleanup_free_ char *pretty = NULL;
+
+                (void) in_addr_to_string(AF_INET, &address, &pretty);
+                log_link_debug(link, "DHCP: Configuring RPDB rule from DHCPv4 address %s", strna(pretty));
+        }
+
+        HASHMAP_FOREACH(rule, link->network->rules_by_section) {
+                if (!IN_SET(rule->to_type, ADDRESS_TYPE_DHCP4_IP, ADDRESS_TYPE_DHCP4_GATEWAY) &&
+                    !IN_SET(rule->from_type, ADDRESS_TYPE_DHCP4_IP, ADDRESS_TYPE_DHCP4_GATEWAY))
+                        continue;
+
+                if ((rule->to_type == ADDRESS_TYPE_DHCP4_GATEWAY || rule->from_type == ADDRESS_TYPE_DHCP4_GATEWAY) && !router) {
+                        log_link_debug(link, "DHCP error: Failed to configure RPDB rule from DHCPv4 gateway");
+                        continue;
+                }
+
+                if (rule->from_type == ADDRESS_TYPE_DHCP4_IP) {
+                        rule->from.in = address.in;
+                        rule->from_prefixlen = prefixlen;
+                }
+
+                if (rule->to_type == ADDRESS_TYPE_DHCP4_IP) {
+                        rule->to.in = address.in;
+                        rule->to_prefixlen = prefixlen;
+                }
+
+                if (rule->from_type == ADDRESS_TYPE_DHCP4_GATEWAY) {
+                        rule->from.in = *router;
+                        rule->from_prefixlen = 32;
+                }
+
+                if (rule->to_type == ADDRESS_TYPE_DHCP4_GATEWAY) {
+                        rule->to.in = *router;
+                        rule->to_prefixlen = 32;
+                }
```

Comment:
Not a strong opinion, but maybe, it is better to use `switch() case:` for readability.